tests/io/datasets/test_hdf5.py

```diff
@@ -423,3 +429,161 @@ def test_hdf5_with_2d_shape(lt_ctx, hdf5_2d):
+    (3, 4, 32, 128),
+    (1, 4, 32, 128),
+    (1, 16, 32, 128),
+])
+def test_chunked_weird(lt_ctx, tmpdir_factory, chunks, udf, shared_random_data):
+    datadir = tmpdir_factory.mktemp('data')
+    filename = os.path.join(datadir, 'weirdly-chunked-256-256.h5')
+    data = shared_random_data
+
+    with h5py.File(filename, "w") as f:
+        f.create_dataset("data", data=data, chunks=chunks)
+
+    ds = lt_ctx.load("hdf5", path=filename)
+
+    p = next(ds.get_partitions())
+    base_shape = p.get_base_shape(roi=None)
+    print(base_shape)
+
+    res = lt_ctx.run_udf(dataset=ds, udf=udf)
+    assert len(res) == 1
+    res = next(iter(res.values()))
+    assert_allclose(
+        res,
+        np.sum(data, axis=(2, 3))
+    )
+
+
+@pytest.mark.parametrize('in_dtype', [
+    np.float32,
+    np.float64,
+    np.uint16,
+])
+@pytest.mark.parametrize('read_dtype', [
+    np.float32,
+    np.float64,
+    np.uint16,
+])
+@pytest.mark.parametrize('use_roi', [
+    True, False
+])
+def test_hdf5_result_dtype(lt_ctx, tmpdir_factory, in_dtype, read_dtype, use_roi):
+    datadir = tmpdir_factory.mktemp('data')
+    filename = os.path.join(datadir, 'result-dtype-checks.h5')
+    data = _mk_random((2, 2, 4, 4), dtype=in_dtype)
+
+    with h5py.File(filename, "w") as f:
+        f.create_dataset("data", data=data)
+
+    ds = lt_ctx.load("hdf5", path=filename)
+
+    if use_roi:
+        roi = np.zeros(ds.shape.nav, dtype=np.bool).reshape((-1,))
+        roi[0] = 1
+    else:
+        roi = None
+    udfs = [SumSigUDF()]  # need to have at least one UDF
+    p = next(ds.get_partitions())
+    neg = Negotiator()
+    tiling_scheme = neg.get_scheme(
+        udfs=udfs,
+        partition=p,
+        read_dtype=read_dtype,
+        roi=roi,
+        corrections=None,
+    )
+    tile = next(p.get_tiles(tiling_scheme=tiling_scheme, roi=roi, dest_dtype=read_dtype))
+    assert tile.dtype == np.dtype(read_dtype)
+
+
+class UDFWithLargeDepth(UDF):
+    def process_tile(self, tile):
+        pass
+
+    def get_tiling_preferences(self):
+        return {
+            "depth": 128,
+            "total_size": UDF.TILE_SIZE_BEST_FIT,
+        }
+
+
+def test_hdf5_tileshape_negotation(lt_ctx, tmpdir_factory):
+    # try to hit the third case in _get_subslices:
+    datadir = tmpdir_factory.mktemp('data')
+    filename = os.path.join(datadir, 'tileshape-neg-test.h5')
+    data = _mk_random((4, 100, 256, 256), dtype=np.uint16)
+
+    with h5py.File(filename, "w") as f:
+        f.create_dataset("data", data=data, chunks=(2, 32, 32, 32))
+
+    ds = lt_ctx.load("hdf5", path=filename)
+
+    udfs = [UDFWithLargeDepth()]
+    p = next(ds.get_partitions())
+    neg = Negotiator()
+    tiling_scheme = neg.get_scheme(
+        udfs=udfs,
+        partition=p,
+        read_dtype=np.float32,
+        roi=None,
+        corrections=None,
+    )
+    assert len(tiling_scheme) > 1
+    next(p.get_tiles(tiling_scheme=tiling_scheme, roi=None, dest_dtype=np.float32))
```

Comment:
It seems that this only tests 4D data?

In terms of testing strategy, what do you think of this approach:

* Test between 2D and, let's say, 7D, with random size of the axes.
* Constant data size. Should be large enough so that a partition doesn't fit the HDF5 tile cache, but small enough so that it fits in the FS cache and main memory.
* Test a random portfolio of chunking schemes, maybe with compression to make sure read amplification "hits"
* Use an UDF that just produces a total sum, go from 1D nav to 1D sig
* Compare the time of this UDF with doing `np.sum(hdf5file["data"])`.

If you like, I can take care of that test suite. :-)